ocaml/xapi/xapi_pool.ml

```diff
@@ -781,15 +781,36 @@ let eject ~__context ~host =
 
 	if Pool_role.is_master () then raise Cannot_eject_master
 	else begin
-		(* Fail the operation if any VMs are running here (except control domains) *)
+		(* Fail the operation if any VMs are running here (except dom0) *)
 		let my_vms_with_records = Db.VM.get_records_where ~__context ~expr:(Eq(Field "resident_on", Literal (Ref.string_of host))) in
 		List.iter (fun (_, x) -> 
-			if (not x.API.vM_is_control_domain) && x.API.vM_power_state<>`Halted
+			if (not (Helpers.is_domain_zero ~__context (Db.VM.get_by_uuid ~__context ~uuid:x.API.vM_uuid))) && x.API.vM_power_state <>`Halted
 			then begin
 				error "VM uuid %s not in Halted state and resident_on this host" (x.API.vM_uuid);
 				raise (Api_errors.Server_error(Api_errors.operation_not_allowed, ["VM resident on host"]))
 			end) my_vms_with_records;
 
+		(* all control domains resident on me should be destroyed once I leave the
+		pool, therefore pick them out as follows: if they have a valid resident_on,
+		the latter should be me; if they don't (e.g. they are halted), they should have
+		disks on my local storage *)
+		let vm_is_resident_on_host vm_rec host =
+			(Db.is_valid_ref __context vm_rec.API.vM_resident_on) && (vm_rec.API.vM_resident_on = host)
+		in
+		let vm_has_disks_on_local_sr_of_host vm_ref host =
```

Comment:
The code has improved a lot. This bit here is still hard to read. How about something like this; im trying once more to make the conditions explicit by giving them names (as you have already). 

		let vm_has_disks_on_local_sr_of_host vm_ref host =
			let non_shared x =  not (Helpers.is_sr_shared ~__context ~self:x) in
			let has_sr x     =  Helpers.check_sr_exists_for_host
				~__context ~self:x ~host:host in
					Db.VM.get_VBDs ~__context ~self:vm_ref
					|> List.map (fun v -> Db.VBD.get_VDI ~__context ~self:v) v
					|> List.filter (fun x -> x <> Ref.null)
					|> List.map (fun v -> Db.VDI.get_SR ~__context ~self:v)
					|> List.exists (fun x -> non_shared x && has_sr x)

It's not clear to me where the `<>` belongs. If you disagree with this, that's fine and I would accept it.